From 5c5ea2e85baf3fb9b24322b338696079d6bd139a Mon Sep 17 00:00:00 2001
From: Andreas Schwab <schwab@linux-m68k.org>
Date: Fri, 25 Jun 2021 15:02:47 +0200
Subject: [PATCH 15/22] wordexp: handle overflow in positional parameter number
 (CVE-2021-35942: bug 28011)

Use strtoul instead of atoi so that overflow can be detected.

Change-Id: I4eeb36a1d58178614f02d43113b2109e30480736
---
 posix/wordexp-test.c | 1 +
 posix/wordexp.c      | 2 +-
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/posix/wordexp-test.c b/posix/wordexp-test.c
index cc29840355..30c1dd65ef 100644
--- a/posix/wordexp-test.c
+++ b/posix/wordexp-test.c
@@ -200,6 +200,7 @@ struct test_case_struct
     { 0, NULL, "$var", 0, 0, { NULL, }, IFS },
     { 0, NULL, "\"\\n\"", 0, 1, { "\\n", }, IFS },
     { 0, NULL, "", 0, 0, { NULL, }, IFS },
+    { 0, NULL, "${1234567890123456789012}", 0, 0, { NULL, }, IFS },
 
     /* Flags not already covered (testit() has special handling for these) */
     { 0, NULL, "one two", WRDE_DOOFFS, 2, { "one", "two", }, IFS },
diff --git a/posix/wordexp.c b/posix/wordexp.c
index 7548e0329f..d43d3e29d7 100644
--- a/posix/wordexp.c
+++ b/posix/wordexp.c
@@ -1419,7 +1419,7 @@ envsubst:
   /* Is it a numeric parameter? */
   else if (isdigit (env[0]))
     {
-      int n = atoi (env);
+      unsigned long n = strtoul (env, NULL, 10);
 
       if (n >= __libc_argc)
 	/* Substitute NULL. */
-- 
2.20.1

